### ISSUE TITLE
[Bug] Qwen3.5/3.8 hybrid GDN: NaN logits after a prefix-cache hit when the previous prefill ended 4-10 tokens past a block boundary (v0.28.0, mamba cache mode align)

### ISSUE THREAD

**Issue by manu-kerrigan** [vllm-project/vllm#55766]
### Your current environment

- vLLM **v0.28.0**, plus a diagnostics-only patch that logs the first NaN per request. Indexing on this path is stock.
- Model: **Qwen/Qwen3.8-27B** BF16 (hybrid Gated-DeltaNet + attention), TP2 on 2x H100-80GB.
- `--enable-prefix-caching`, `--speculative-config '{"method":"ngram","num_speculative_tokens":5,"prompt_lookup_max":5,"prompt_lookup_min":2}'`, `--reasoning-parser qwen3 --tool-call-parser qwen3_coder --enable-auto-tool-choice`, `--max-model-len 131072`, chunked prefill with `max_num_batched_tokens 8192`.
- Mamba cache mode is `align` (the default when prefix caching is on; `all` is unavailable for this model). The engine sets the attention block size to **816 tokens** to match the mamba page size.

### Describe the bug

A request returns logits that are NaN across the whole vocabulary from its first step, so it samples token id 0 (`"!"`) until `max_tokens` — 32,768 tokens over 11 minutes in the incident that found this. `vllm:corrupted_requests_total` increments. Retrying the identical request fails identically until the prefix cache turns over.

### Reproduction (content-free, two requests, about 30 seconds)

1. **Request A**: one user message of random words whose prompt length is exactly `N*816 + r` tokens, with `r` in {6, 8}. `r` in {4, 10} reproduces in about half the trials; `r <= 3`, `r >= 12` and every odd `r` were clean. Verified at N = 10, 20, 34, 40. temperature 1.0, top_p 0.95, top_k 20, max_tokens 300, thinking on.
2. **Request B**: A's messages + A's reply as an assistant message (including its reasoning) + a further ~2,000-token user message, max_tokens 64. B gets a block-aligned prefix hit up to `N*816`, prefills from there, and its logits are NaN from the first step.

Controls that isolate the carrier:

| variant of B | result |
| --- | --- |
| unchanged | NaN |
| with a per-request `cache_salt` | clean |
| 12 pad tokens **prepended** to A's content (whole prefix shifts, no cache hit) | clean |
| 12 pad tokens **appended** to A's content (hit unchanged) | NaN |
| short B (~16 new tokens) | clean — B must prefill at least a full block after the restored boundary |

Reproduced on a cold engine and on one that had been serving for hours.

### What the engine reports at the failing step

```
NAN_LOGITS first NaN for request ...: num_computed_tokens=27744 num_scheduled_tokens=1632
  num_drafts=0 accepted_in=1 mamba_col=35 mamba_block=691 batch_size=1 reqs_with_drafts=0
  logits_rows=1
```

That is the first uncached chunk after a 34-block hit, with no drafts in the batch, a single-request batch, and `vllm:num_preemptions_total` at 0.

### Analysis so far (root cause not established)

Decode-produced boundary checkpoints are never prefix-cached in v0.28.0 (the running block is freed before `cache_blocks`), so only prefill chunk ends matter, which is consistent with the trigger being the prompt length. The Mamba/GDN checkpoint written when a prefill ends 4 to 10 tokens past a block boundary appears to be bad; a later request restores it from the prefix cache, produces NaN, and the NaN block is then itself cached, which is why the retry fails identically. Speculation was not active at the failing step, though whether the ngram path is required upstream of it is untested. `--mamba-cache-mode all` is not available for this model and silently falls back to `align`.

### Impact

Agent tool loops grow a conversation by hundreds of tokens per call, so each call is a fresh draw of prompt length mod 816. We observed empty replies, 32k-token `"!"` outputs, and identical retries failing identically.

### Workarounds (client-side only; we have no engine fix)

- Salt the next request (`cache_salt`) when the previous prompt's length mod 816 falls in {4, 6, 8, 10}.
- On a retry, pad the **start** of the prompt. A pad appended at the end does not help.

### Possibly related

#39273 (GDN state corruption with ngram speculation), #53912 ("!!!!" / empty output on Qwen3.5-27B with MTP on v0.28.0), #54173, #55506, #45238.

### Before submitting a new issue...

- [x] I searched the existing issues and did not find this one.
- [x] I am reporting a reproducible bug with the information above.

**Comment by JulianZJN** [comment]
I'd like to investigate this, starting with a version-controlled reproduction rather than an assumed fix. I checked the issue discussion and PR references; I haven't found a PR explicitly targeting this report.

I'll compare v0.28.0 with current main, sweep producer lengths around the resolved Mamba block boundary, and test ngram on/off plus cold/salted controls. The next step is to distinguish a bad checkpoint at write time from a wrong restore slot or a failure during the first uncached suffix prefill. I'll check the adjacent GDN state-correction and Mamba alignment work before proposing a patch.

I have SM89 hardware available, so a smaller same-family model or TP4 reproduction will be an additional configuration, not a claim to have reproduced your TP2 H100 setup. Do you have a shareable synthetic request-pair generator (including how the assistant reasoning is carried into request B)? That would help preserve the original trigger while avoiding private conversation data.

**Comment by JulianZJN** [comment]
Follow-up on my earlier comment. On 4x RTX 4090 (SM89, TP4) I can reproduce all-NaN logits with `Qwen/Qwen3.8-27B` BF16, prefix caching on, `mamba_cache_mode=align` and ngram `num_speculative_tokens=5` (block size resolves to 816), on both `v0.28.0` and main `71fc70d3`. In these runs the trigger is a final prefill chunk of exactly `1 + num_speculative_tokens` = 6 tokens, which is the graph-dispatch problem described in #49918 and addressed by the runner change in #47123. It is not the align-mode chunk splitting in #54076.

This does not explain the whole report yet: I could not reproduce `r = 8`, and in my failing pairs A is already corrupted, while here A was fine. Details and limits below.

What happens: With `mamba_cache_mode=align` the prefill is split on the 816-token grid, so a prompt of `N*816 + 6` tokens ends in a 6-token chunk, which equals `1 + num_speculative_tokens`. `_is_uniform_decode` (`gpu_model_runner.py`) only looks at the shape, so that prefill step is dispatched to the captured FULL spec-verify decode graph. The GDN metadata builder treats the same step as a prefill and skips refreshing its persistent spec-decode buffers (`gdn_attn.py`, the `num_prefills == 0` branches), so the graph replays whatever the previous spec-decode step left there: the previous request's state block ids and `num_accepted_tokens`.

Tracing per-block checksums of the layer-0 GDN state and a finiteness scan of the first attention layer's KV around each step:

- on that 6-token step the blocks written are exactly the block set of the previous request's last spec step; none of them is referenced by the current step's metadata
- one of them had meanwhile been reallocated as the running request's first full-attention block (the GDN layers and the attention layers share the KV cache tensor); the fp32 SSM state written into it reads back as non-finite bf16 KV in the same forward, and the request returns NaN logits
- that block is a full block, so it is registered in the prefix cache. The follow-up request B restores it (16320-token hit) and NaNs on its first step; B with a new `cache_salt` gets fresh blocks and is clean. That matches the salt / shift / "persists until evicted" behaviour reported here
- with `cudagraph_mode=PIECEWISE` the same step writes only its own three state blocks (traced step by step). With `enforce_eager` or ngram off the workload is clean, which I checked by corrupted-request count only. On a freshly started server the very first colliding request is not NaN either (the stale buffers still hold the capture-time null block, so the state write is skipped, which is the #49918 symptom); it starts once an earlier request has done a full-draft spec step

A->B matrix on main, 3 seeds per cell, fresh server per cell (requests with NaN logits, shown as A, B, B with new `cache_salt`):

| r | main | main + #47123 |
|---|---|---|
| 6 (N=20) | 2/3, 2/3, 0/3 | 0/3, 0/3, 0/3 |
| 6 (N=34) | 2/3, 2/3, 0/3 | 0/3, 0/3, 0/3 |
| 8 (N=20, 34) | 0/3, 0/3, 0/3 | 0/3, 0/3, 0/3 |
| 4, 7, 10 | 0/3, 0/3, 0/3 | 0/3, 0/3, 0/3 |

In that table each arm built B from its own reply to A, so the two arms did not see identical B inputs (when A fails it produces no reply). A rerun with frozen inputs, the same token ids sent to both builds (`r = 6`, `N = 20`, 3 seeds, fresh server per arm), gives the same result: 2/3, 2/3, 0/3 on main and 0/3, 0/3, 0/3 with the runner change.

With #47123's runner change the 6-token tail runs PIECEWISE, a 10-request `r=6` workload goes from 5–9 NaN requests to 0 (two server starts), and the state checksums on the colliding step are identical to the PIECEWISE run.

What this does not explain: Here only `r = K + 1 = 6` triggers it; `r = 8` (and 4 / 10) are clean, and when B fails, A has failed too, because the clobbered attention block is A's own. In the original report A was fine and `r = 8` was as reliable as 6. @manu-kerrigan if you still have the setup, the scheduler log line for A's last prefill step (`num_scheduled_tokens`) would tell us whether your tail chunk was also 6 tokens for the `r = 8` prompts (e.g. if the token count was taken before the chat template). I'm on SM89, so I also can't rule out an SM90-specific difference in which blocks get hit.

Workaround until #47123 lands: `--compilation-config '{"cudagraph_mode": "PIECEWISE"}'`. Setting `num_speculative_tokens` to another value only moves the colliding residue (with `1` it is `r = 2`, which NaNs the same way).

Drivers, per-request result tables with input hashes, the applied diff, server command lines and the traces are in https://github.com/JulianZJN/vllm-55766-evidence.
